### DISCUSSION TITLE
Mixins

### DISCUSSION THREAD

**Discussion by alimd** [AliMD/alwatr#87] [Educational]
## What is a Mixin?

> A mixin is an abstract subclass; i.e. a subclass definition that may be applied to different superclasses to create a related family of modified classes.
>
> - Gilad Bracha and William Cook, [Mixin-based Inheritance](http://www.bracha.org/oopsla90.pdf)

Let's take for example Logging. Imagine you have 3 Pages

- Red
- Green
- Blue

```
              +----------+
              |   Page   |
              +----------+
                |  |  |
     +----------+  |  +-----------+
     |             |              |
+---------+ +-----------+ +----------+
| PageRed | | PageGreen | | PageBlue |
+----+----+ +-----------+ +----------+
```

```js
class Page {}
class PageRed extends Page {}
class PageGreen extends Page {}
class PageBlue extends Page {}
```

Now we want to log whenever someone goes on Page Red.
To archive that we extend Page Red and make a Logged Page Red.

```
              +----------+
              |   Page   |
              +-+--+--+--+
                |  |  |
     +----------+  |  +-----------+
     |             |              |
+----+----+  +-----+-----+  +-----+----+
| PageRed |  | PageGreen |  | PageBlue |
+----+----+  +-----------+  +----------+
     |
+----+----+
| Logged  |
| PageRed |
+---------+
```

```js
class Page {}
class PageRed extends Page {}
class PageGreen extends Page {}
class PageBlue extends Page {}
class LoggedPagRed extends PageRed {}
```

If we want to start logging for PageGreen we have an issue:

- we can't put the logic in `Page` as Blue should not be logged
- we can't reuse the logic in `Logged PageGreen` as we can not extend from 2 source (even if we could it would mean conflicting info in Red and Green)

What we can do is put it in an "external" place and write it so it can be "mixed in".

```
               +----------+                +----------+
               |   Page   |                | Logging* |
               +-+--+--+--+                +----------+
                 |  |  |
      +----------+  |  +-----------+
      |             |              |
+-----+----+  +-----+-----+  +-----+----+
| PageRed  |  | PageGreen |  | PageBlue |
|  with    |  |   with    |  +----------+
| Logging* |  |  Logging* |
+----------+  +-----------+
```

```js
// defining the Mixin
export const LoggingMixin = superclass =>
  class LoggingMixin extends superclass {
    // logging logic
  };
class Page {}
// applying a Mixin
class PageRed extends LoggingMixin(Page) {}
class PageGreen extends LoggingMixin(Page) {}
class PageBlue extends Page {}
```

With that approach we can extract logic into a separate code pieces we can use where needed.

### More in depth explanation

- [Class mixins](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Classes#mix-ins) on MDN
- [Real Mixins with JavaScript Classes](https://justinfagnani.com/2015/12/21/real-mixins-with-JavaScript-classes/) by Justin Fagnani
- [Mixins](https://www.typescriptlang.org/docs/handbook/mixins.html) in the TypeScript handbook.
- [Dedupe mixin library](https://open-wc.org/docs/development/dedupe-mixin/) by open-wc, including a discussion of when mixin usage may lead to duplication, and how to use a deduping library to avoid it.
- [Mixin conventions](https://component.kitchen/elix/mixins) followed by Elix web component library. While not Lit-specific, contains thoughtful suggestions around applying conventions when defining mixins for web components.
- [Creating mixins for LitElement](https://lit.dev/docs/composition/mixins/#creating-mixins-for-litelement)

**Comment by asmanp2012** [comment]
https://javascript.info/mixins#a-mixin-example
